File: adapters/python/check_refs.py
#!/usr/bin/env python3
"""扫描 AGENTS.md / README.md / docs/**.md 中的路径引用，报告不存在的路径。

Python adapter，对齐 adapters/node/check_refs.mjs 的 path-ref 部分。
不实现项目特异的 CLI 命令扫描（universal contract 之外，由下游 lane 自加）。
共用 agents-md.config.json 真源；仅依赖 Python 3 标准库。
"""

from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path

ROOT = Path(os.environ["AGENTS_MD_ROOT"]).resolve() if os.environ.get("AGENTS_MD_ROOT") else Path(__file__).resolve().parents[2]
CONFIG = json.loads((ROOT / "agents-md.config.json").read_text(encoding="utf-8"))
KNOWN_PREFIXES: list[str] = CONFIG["knownPrefixes"]
RUNTIME_PREFIXES: list[str] = CONFIG["runtimePrefixes"]
GENERATED_PREFIXES: list[str] = CONFIG["generatedPrefixes"]
INTENTIONALLY_ABSENT_REFS: set[str] = set(CONFIG["intentionallyAbsentRefs"])
SCAN_EXCLUDE_DIRS: list[str] = CONFIG.get("scanExcludeDirs", [])

BACKTICK_RE = re.compile(r"`([^`]+)`")
LINK_RE = re.compile(r"\[[^\]]*\]\(([^)\s]+)\)")
SCHEME_RE = re.compile(r"^[a-z][a-z0-9+.\-]*:", re.IGNORECASE)


def is_path_ref(ref: str) -> bool:
    if "/" not in ref:
        return False
    if "*" in ref or "<" in ref:
        return False
    if " " in ref:
        return False
    return any(ref.startswith(p) for p in KNOWN_PREFIXES)


def normalize_link(source_dir: Path, target: str) -> str | None:
    hash_idx = target.find("#")
    path = target[:hash_idx] if hash_idx >= 0 else target
    if not path:
        return None
    if SCHEME_RE.match(path):
        return None
    if path.startswith("/"):
        return None
    abs_path = (source_dir / path).resolve()
    try:
        rel = abs_path.relative_to(ROOT)
    except ValueError:
        return None
    return rel.as_posix()


def collect_markdown_files() -> list[Path]:
    files: list[Path] = [ROOT / "AGENTS.md", ROOT / "README.md"]
    docs_dir = ROOT / "docs"
    if docs_dir.exists():
        for path in sorted(docs_dir.rglob("*.md")):
            rel = path.relative_to(ROOT).as_posix()
            if any(rel == p or rel.startswith(p + "/") for p in SCAN_EXCLUDE_DIRS):
                continue
            # Also skip if any parent dir is in exclude list
            parent_rel = path.parent.relative_to(ROOT).as_posix()
            if any(parent_rel == p or parent_rel.startswith(p + "/") for p in SCAN_EXCLUDE_DIRS):
                continue
            files.append(path)
    return [f for f in files if f.exists() and f.is_file()]


def extract_refs(file_path: Path) -> list[tuple[str, str]]:
    md = file_path.read_text(encoding="utf-8")
    source = file_path.relative_to(ROOT).as_posix()
    source_dir = file_path.parent
    refs: list[tuple[str, str]] = []
    for match in BACKTICK_RE.finditer(md):
        ref = match.group(1)
        if is_path_ref(ref):
            refs.append((ref, source))
    for match in LINK_RE.finditer(md):
        normalized = normalize_link(source_dir, match.group(1))
        if normalized is not None:
            refs.append((normalized, source))
    return refs


def should_check_existence(path: str) -> bool:
    if any(path.startswith(p) for p in RUNTIME_PREFIXES):
        return False
    if any(path.startswith(p) for p in GENERATED_PREFIXES):
        return False
    if path in INTENTIONALLY_ABSENT_REFS:
        return False
    return True


def main() -> int:
    unique: dict[str, str] = {}
    for file_path in collect_markdown_files():
        for path, source in extract_refs(file_path):
            unique.setdefault(path, source)

    dead = [
        (path, source)
        for path, source in unique.items()
        if should_check_existence(path) and not (ROOT / path).exists()
    ]

    if dead:
        print(f"Markdown reference check FAILED — {len(dead)} dead ref(s):\n", file=sys.stderr)
        for path, source in dead:
            print(f"  DEAD: {path} (referenced by {source})", file=sys.stderr)
        return 1
    print(f"Markdown reference check PASSED — {len(unique)} path refs checked, 0 dead.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: adapters/node/check_refs.mjs
#!/usr/bin/env node
/**
 * 扫描 scanMarkdownRoots 配置的 .md 文件 / 目录中反引号内与 Markdown link 内的文件路径引用，报告不存在的路径。
 * 只检查含 "/" 的引用（完整路径），跳过裸文件名、glob 模式和模板占位符。
 */
import { readFileSync, existsSync, readdirSync, statSync } from 'node:fs';
import { dirname, relative, resolve } from 'node:path';
import { pathToFileURL } from 'node:url';

const REPO_ROOT = process.env.AGENTS_MD_ROOT
  ? resolve(process.env.AGENTS_MD_ROOT)
  : resolve(import.meta.dirname, '../..');
const CONFIG = JSON.parse(
  readFileSync(resolve(REPO_ROOT, 'agents-md.config.json'), 'utf8'),
);
const KNOWN_PREFIXES = CONFIG.knownPrefixes ?? [];
const RUNTIME_PREFIXES = CONFIG.runtimePrefixes ?? [];
const GENERATED_PREFIXES = CONFIG.generatedPrefixes ?? [];
const INTENTIONALLY_ABSENT_REFS = new Set(CONFIG.intentionallyAbsentRefs ?? []);
const SCAN_EXCLUDE_DIRS = CONFIG.scanExcludeDirs ?? [];
const SCAN_MARKDOWN_ROOTS = CONFIG.scanMarkdownRoots ?? ['AGENTS.md', 'README.md', 'docs/'];

function toUnixPath(path) {
  return path.split('\\').join('/');
}

function collectMarkdownFiles(root, dir) {
  if (!existsSync(dir)) return [];
  const relPath = toUnixPath(relative(root, dir));
  if (SCAN_EXCLUDE_DIRS.some((p) => relPath === p || relPath.startsWith(`${p}/`))) return [];
  const files = [];
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const fullPath = resolve(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...collectMarkdownFiles(root, fullPath));
    } else if (entry.isFile() && entry.name.endsWith('.md')) {
      files.push(fullPath);
    }
  }
  return files;
}

/**
 * KNOWN_PREFIXES 用正向匹配排除 CLI 命令 / npm 包名 / 路由 / slash command 等误报；
 * RUNTIME_PREFIXES / GENERATED_PREFIXES / INTENTIONALLY_ABSENT_REFS 跳过存在性校验。
 * 真源在 agents-md.config.json。
 */

function isPathRef(ref) {
  if (!ref.includes('/')) return false;
  if (ref.includes('*') || ref.includes('<')) return false;
  if (ref.includes(' ')) return false; // CLI 命令含空格
  return KNOWN_PREFIXES.some((p) => ref.startsWith(p));
}

function extractRefs(root, filePath) {
  const md = readFileSync(filePath, 'utf8');
  const source = toUnixPath(relative(root, filePath));
  const sourceDir = dirname(filePath);
  const refs = [];

  for (const match of md.matchAll(/`([^`]+)`/g)) {
    const ref = match[1];
    if (isPathRef(ref)) refs.push({ path: ref, source });
  }

  for (const match of md.matchAll(/\[[^\]]*\]\(([^)\s]+)\)/g)) {
    const normalized = normalizeMarkdownLink(root, sourceDir, match[1]);
    if (normalized) refs.push({ path: normalized, source });
  }

  return refs;
}

function normalizeMarkdownLink(root, sourceDir, target) {
  const hashIdx = target.indexOf('#');
  const path = hashIdx >= 0 ? target.slice(0, hashIdx) : target;
  if (!path) return null;
  if (/^[a-z][a-z0-9+.-]*:/i.test(path)) return null;
  if (path.startsWith('/')) return null;
  const rel = toUnixPath(relative(root, resolve(sourceDir, path)));
  if (!rel || rel.startsWith('..')) return null;
  return rel;
}

function shouldCheckExistence(path) {
  if (RUNTIME_PREFIXES.some((prefix) => path.startsWith(prefix))) return false;
  if (GENERATED_PREFIXES.some((prefix) => path.startsWith(prefix))) return false;
  if (INTENTIONALLY_ABSENT_REFS.has(path)) return false;
  return true;
}

function resolveScanTargets(root, scanRoots) {
  const targets = [];
  for (const entry of scanRoots) {
    const abs = resolve(root, entry);
    if (!existsSync(abs)) continue;
    const st = statSync(abs);
    if (st.isDirectory()) {
      targets.push(...collectMarkdownFiles(root, abs));
    } else if (st.isFile() && entry.endsWith('.md')) {
      targets.push(abs);
    }
  }
  return targets;
}

export function checkMarkdownRefs(root = process.cwd()) {
  const markdownFiles = resolveScanTargets(root, SCAN_MARKDOWN_ROOTS);

  /** 去重，保留每个路径最早出现的位置 */
  const unique = new Map();
  for (const ref of markdownFiles.flatMap((filePath) => extractRefs(root, filePath))) {
    if (!unique.has(ref.path)) unique.set(ref.path, ref.source);
  }

  /** 运行态和构建产物路径只检查形态，不要求本地文件存在 */
  const dead = [...unique.entries()].filter(([path]) => {
    if (!shouldCheckExistence(path)) return false;
    return !existsSync(resolve(root, path));
  });

  return {
    checkedCount: unique.size,
    dead: dead.map(([path, source]) => ({ path, source })),
  };
}

function runCli() {
  const result = checkMarkdownRefs(REPO_ROOT);
  if (result.dead.length > 0) {
    console.error(`Markdown reference check FAILED — ${result.dead.length} dead ref(s):\n`);
    for (const { path, source } of result.dead) {
      console.error(`  DEAD: ${path} (referenced by ${source})`);
    }
    process.exit(1);
  }

  console.log(`Markdown reference check PASSED — ${result.checkedCount} path refs, 0 dead.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  runCli();
}
